src/core/sync/mutex.d

```diff
@@ -125,61 +156,75 @@ class Mutex :
      * If this lock is not already held by the caller, the lock is acquired,
      * then the internal counter is incremented by one.
      *
-     * Throws:
-     *  SyncError on error.
+     * Note:
+     *    `Mutex.lock` does not throw, but a class derived from Mutex can throw.
+     *    Use `lock_nothrow` in `nothrow @nogc` code.
      */
     @trusted void lock()
     {
         lock_nothrow();
     }
 
-    // undocumented function for internal use
-    final void lock_nothrow() nothrow @trusted @nogc
+    /// ditto
+    @trusted void lock() shared
+    {
+        lock_nothrow();
+    }
+
+    /// ditto
+    final void lock_nothrow(this Q)() nothrow @trusted @nogc
+        if (is(Q == Mutex) || is(Q == shared Mutex))
     {
-        version( Windows )
+        version (Windows)
         {
-            EnterCriticalSection( &m_hndl );
+            EnterCriticalSection(&m_hndl);
         }
-        else version( Posix )
+        else version (Posix)
         {
-            int rc = pthread_mutex_lock( &m_hndl );
-            if( rc )
-            {
-                SyncError syncErr = cast(SyncError) cast(void*) typeid(SyncError).initializer;
-                syncErr.msg = "Unable to lock mutex.";
-                throw syncErr;
-            }
+            if (pthread_mutex_lock(&m_hndl) == 0)
+                return;
+
+            SyncError syncErr = cast(SyncError) cast(void*) typeid(SyncError).initializer;
+            syncErr.msg = "Unable to lock mutex.";
+            throw syncErr;
         }
     }
 
     /**
      * Decrements the internal lock count by one.  If this brings the count to
      * zero, the lock is released.
      *
-     * Throws:
-     *  SyncError on error.
+     * Note:
+     *    `Mutex.unlock` does not throw, but a class derived from Mutex can throw.
+     *    Use `unlock_nothrow` in `nothrow @nogc` code.
      */
     @trusted void unlock()
     {
         unlock_nothrow();
     }
 
-    // undocumented function for internal use
-    final void unlock_nothrow() nothrow @trusted @nogc
+    /// ditto
+    @trusted void unlock() shared
+    {
+        unlock_nothrow();
+    }
+
+    /// ditto
+    final void unlock_nothrow(this Q)() nothrow @trusted @nogc
```

Comment:
`unlock_nothrow` was introduced in 2.067, so there's not much that I can do now.